### PR TITLE
Complete azure-identity prototype

### LINKED ISSUE #5149: Credential classes
- [ ] ~~AzureCredential~~
- [x] Sync
  - [x] ClientSecretCredential
  - [x] ClientCertificateCredential
  - [x] MSICredential
- [x] Async
  - [x] ClientSecretCredential
  - [x] ClientCertificateCredential
  - [x] MSICredential

### LINKED ISSUE #5150: Credential provider classes
- [ ] ~~TokenCredentialProvider~~
- [x] TokenCredentialProviderChain
- [x] EnvironmentCredentialProvider
- [x] MSICredentialProvider

### LINKED ISSUE #5251: Implement certificate credential provider

### PR DESCRIPTION
**Pull request by chlowell**
This implements the rest of the azure-identity prototype introduced by #5246, adding the remaining credentials with unit tests. With this the package supports sync and async AAD authentication with managed identity or with a certificate or client secret. MSAL is used for token caching and signing the JWT assertion used in cert authentication.

Future work for future PRs includes more documentation, live tests, and maybe more shared code between credentials and between a/sync equivalents. Suggestions around the latter are welcome. Sharing more code is possible but I found everything I tried too difficult to read.

Closes #5251, closes #5149, closes #5150

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/identity/azure-identity/azure/identity/_base.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+class ClientSecretCredentialBase(object):
+    def __init__(self, client_id, secret, tenant_id, **kwargs):
+        # type: (str, str, str, Mapping[str, Any]) -> None
+        if not client_id:
+            raise ValueError("client_id")
+        if not secret:
+            raise ValueError("secret")
+        if not tenant_id:
+            raise ValueError("tenant_id")
```
Comment (bryevdv) [inline sdk/identity/azure-identity/azure/identity/_base.py]: The msgs to `ValueError` could be a little more informative

File: sdk/identity/azure-identity/azure/identity/aio/credentials.py [hunk 2]
```diff
@@ -3,64 +3,130 @@
...
+            )
+        return token  # type: ignore
+
+    @staticmethod
+    def create_config(**kwargs: Mapping[str, Any]) -> Configuration:
+        config = Configuration(**kwargs)
+        config.header_policy = HeadersPolicy(base_headers={"Metadata": "true"}, **kwargs)
+        config.logging_policy = NetworkTraceLoggingPolicy(**kwargs)
+        config.retry_policy = AsyncRetryPolicy(
+            retry_on_status_codes=[404, 429] + [x for x in range(500, 600)], **kwargs
```
Comment (bryevdv) [inline sdk/identity/azure-identity/azure/identity/aio/credentials.py]: `list(range(500, 600))` also works

File: sdk/identity/azure-identity/azure/identity/version.py [hunk 3]
```diff
@@ -3,4 +3,4 @@
 # Licensed under the MIT License. See LICENSE.txt in the project root for
 # license information.
 # --------------------------------------------------------------------------
-VERSION = "0.0.1"
+VERSION = "1.0.0"
```
Comment (schaabs) [inline sdk/identity/azure-identity/azure/identity/version.py]: This should be a prerelease so 1.0.0b1 possibly?

File: sdk/identity/azure-identity/azure/identity/_base.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+class ClientSecretCredentialBase(object):
+    def __init__(self, client_id, secret, tenant_id, **kwargs):
+        # type: (str, str, str, Mapping[str, Any]) -> None
+        if not client_id:
+            raise ValueError("client_id should be the id of an Azure Active Directory application")
+        if not secret:
+            raise ValueError("secret should be an Azure Active Directory application's client secret")
+        if not tenant_id:
+            raise ValueError("tenant_id should be an Azure Active Directory tenant's id (also called its 'directory' id)")
```
Comment (johanste) [inline sdk/identity/azure-identity/azure/identity/_base.py]: Is it also called 'directory' or 'directory id'? The quote placement confuses me.

File: sdk/identity/azure-identity/azure/identity/aio/_authn_client.py [hunk 5]
```diff
@@ -48,6 +49,6 @@ def create_config(**kwargs: Mapping[str, Any]) -> Configuration:
         config = Configuration(**kwargs)
         config.logging_policy = NetworkTraceLoggingPolicy(**kwargs)
         config.retry_policy = AsyncRetryPolicy(
-            retry_on_status_codes=[404, 429] + [x for x in range(500, 600)], **kwargs
+            retry_on_status_codes=[404, 429] + list(range(500, 600)), **kwargs
```
Comment (schaabs) [inline sdk/identity/azure-identity/azure/identity/aio/_authn_client.py]: >404 [](start = 35, length = 3)

Why are you retrying on 404?

File: sdk/identity/azure-identity/azure/identity/constants.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class EnvironmentVariables:
+    # TODO: align cross-language
+    AZURE_CLIENT_ID = "AZURE_CLIENT_ID"
+    AZURE_CLIENT_SECRET = "AZURE_CLIENT_SECRET"
+    AZURE_TENANT_ID = "AZURE_TENANT_ID"
+    CLIENT_SECRET_VARS = (AZURE_CLIENT_ID, AZURE_CLIENT_SECRET, AZURE_TENANT_ID)
+
+    AZURE_PRIVATE_KEY_FILE = "AZURE_PRIVATE_KEY_FILE"
```
Comment (schaabs) [inline sdk/identity/azure-identity/azure/identity/constants.py]: >AZURE_PRIVATE_KEY_FILE [](start = 30, length = 22)

AZURE_CLIENT_CERTIFICATE_PATH

File: sdk/identity/azure-identity/azure/identity/constants.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    AZURE_THUMBPRINT = "AZURE_THUMBPRINT"
```
Comment (schaabs) [inline sdk/identity/azure-identity/azure/identity/constants.py]: >AZURE_THUMBPRINT [](start = 24, length = 16)

AZURE_CLIENT_CERTIFICATE_THUMBPRINT

File: sdk/identity/azure-identity/azure/identity/credentials.py [hunk 8]
```diff
@@ -15,42 +20,115 @@
...
-    def __init__(self, client_id, secret, tenant_id, config=None, **kwargs):
-        # type: (str, str, str, Optional[Configuration], Mapping[str, Any]) -> None
-        if not secret:
-            raise ValueError("secret")
-        super(ClientSecretCredential, self).__init__(client_id, tenant_id, config, **kwargs)
-        self._form_data = {"client_id": client_id, "client_secret": secret, "grant_type": "client_credentials"}
+class CertificateCredential(CertificateCredentialBase):
+    """Authenticates with a certificate (thumbprint and PEM-encoded private key)"""
+
+    def __init__(self, client_id, tenant_id, private_key, thumbprint, config=None, **kwargs):
```
Comment (schaabs) [inline sdk/identity/azure-identity/azure/identity/credentials.py]: >def __init__(self, client_id, tenant_id, private_key, thumbprint, config=None, **kwargs): [](start = 4, length = 89)

private key should be the certificate path. Also the path and thumbprint would both be optional (but you have to specify one).  That being said the usage of thumbprint is that I can find the cert with that thumbprint in the cert store, which we're not supporting yet so perhaps just remove thumbprint for now?

File: sdk/identity/azure-identity/azure/identity/aio/_authn_client.py (hunk 5 above)
Comment (chlowell) [reply sdk/identity/azure-identity/azure/identity/aio/_authn_client.py]: Right? According to the [IMDS docs](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#error-handling) the endpoint returns 404 when it's updating, and clients should retry.

Comment (johanste) [reply sdk/identity/azure-identity/azure/identity/aio/_authn_client.py]: I'm not a big fan of having to create a list with >100 items. There are some things we can do to improve. Just off the top of my head, something like:

```python
class StatusCodeRange(object):

    def __init__(self, min_value, max_value):
        self.min_value = min_value
        self.max_value = max_value

    def __eq__(self, other):
        return self.min_value <= other and other <= self.max_value


scr = StatusCodeRange(500, 600)

assert 503 in [404, 429, scr]
assert 499 not in [1, 2, scr]
```
It does bastardize the __eq__ method, though, so another version would be to change the simple `in` check in the retry policy to understand ranges as well...

File: sdk/identity/azure-identity/azure/identity/version.py (hunk 3 above)
Comment (johanste) [reply sdk/identity/azure-identity/azure/identity/version.py]: Yes. We should mark this as a preview release...

File: sdk/identity/azure-identity/azure/identity/aio/_authn_client.py (hunk 5 above)
Comment (johanste) [reply sdk/identity/azure-identity/azure/identity/aio/_authn_client.py]: > Right? According to the [IMDS docs](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#error-handling) the endpoint returns 404 when it's updating, and clients should retry.

Per the linked documentation (great find, btw :)), we should make sure our retry policy follows the suggested timeouts by default. We are trying to make sure that our libraries "do the right thing" (tm).

Comment (chlowell) [reply sdk/identity/azure-identity/azure/identity/aio/_authn_client.py]: Have done. I put #5628 on the backlog to track retry code range support.

**Comment by chlowell** [comment]
I plan to merge this once CI is green and open another PR to incorporate the architecture board's feedback.

**Pull request merged**